sensor: icm45686: Change watermark threshold irq to equals

anthony289478 opened the pull request:
When using the default configuration of watermark threshold interrupt greater than or equals, extra interrupts are serviced to icm45686_event_handler(). The expected behavior should be only one interrupt is generated per watermark threshold crossing, and until the host drains the fifo, no extra interrupts should be generated.

IMU dt fragment used,
```
&icm45686 {
	status = "okay";
	accel-odr = <ICM45686_DT_ACCEL_ODR_1600>;
	accel-fs = <ICM45686_DT_ACCEL_FS_16>;
	accel-pwr-mode = <ICM45686_DT_ACCEL_LN>;
	gyro-odr = <ICM45686_DT_GYRO_ODR_1600>;
	gyro-fs = <ICM45686_DT_GYRO_FS_2000>;
	gyro-pwr-mode = <ICM45686_DT_GYRO_LN>;
	fifo-watermark = <64>;
};
```
This is the observed current behavior of the interrupt pin with the default setting, per datasheet.
```
1: Write watermark interrupt generated when FIFO data count is greater
than or equal to FIFO watermark threshold
```
<img width="1750" height="964" alt="image" src="https://github.com/user-attachments/assets/58aae626-8757-4fd0-a81e-f69049e7cd67" />


\
This is the proposed expected behavior of the interrupt pin with the new setting, per datasheet.
```
0: Write watermark interrupt generated when FIFO data count is equal to
the FIFO watermark threshold
```
<img width="1753" height="964" alt="image" src="https://github.com/user-attachments/assets/5d011c57-fecf-463f-9d6d-29ae2c87bbc7" />

ubieda commented:
This change would mute warnings thrown due to system latency on servicing the FIFO Watermark events, wouldn't it?

anthony289478 commented:
> This change would mute warnings thrown due to system latency on servicing the FIFO Watermark events, wouldn't it?

Could you explain a bit more? From what I see, the rtio submission occurs in gpio interrupt context so it is always serviced.
The main issue for me is when there is system latency and the rtio workq does not immediately drain the fifo extra interrupts are generated which allocate more rtio transactions from the context mempool.

anthony289478 commented:
> Right - so there's no good reason to have the Workqueue used for every submission. 

You still do since `icm45686_prep_reg_read_rtio_async()` and others still creates async queue's and `rtio_submit` is called in `icm45686_event_handler()`
sensor rtio submissions by default are processed [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/default_rtio_sensor.c#L286) via the `rtio_workq`

>In this case, subsequent FIFO_WM triggers are ignored at the handler level although logged as LOG_WRN. 

I see you've added [disabling interrupts](https://github.com/zephyrproject-rtos/zephyr/pull/94832/files#diff-a9718e5a998674de0d95c1a5f255169848ee689f88f2b123d47ca258b21fd16bR187-R188) in icm45686_event_handler but to me that is a software fix for improperly configured hardware.
Could you elaborate based on that PR, without disabling interrupts and configuring wm on equals only, how the system would drop interrupts due to system latency?

From what I understand, once triggered by hardware, a vectored interrupt, even when preempted by higher priority interrupts, will always be serviced by the system eventually so `icm45686_gpio_callback()` and thus `icm45686_event_handler()` is guaranteed to be called on every edge when interrupts are enabled.

ubieda commented:
> > Right - so there's no good reason to have the Workqueue used for every submission.
> 
> You still do since `icm45686_prep_reg_read_rtio_async()` and others still creates async queue's and `rtio_submit` is called in `icm45686_event_handler()` sensor rtio submissions by default are processed [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/default_rtio_sensor.c#L286) via the `rtio_workq`

The sensor_submit_fallback only kicks-in if the sensor drivers doesn't implement the `.submit` API, which in this case it does:
https://github.com/zephyrproject-rtos/zephyr/blob/d494f86ea40ef470ddaa2bb15ae83234345e7216/drivers/sensor/default_rtio_sensor.c#L32-L38

> Could you elaborate based on that PR, without disabling interrupts and configuring wm on equals only, how the system would drop interrupts due to system latency?

Take a look at:
https://github.com/CogniPilot/zephyr/blob/a03c53eea78e4bb7c050177a938d214c2f6273d0/drivers/sensor/tdk/icm45686/icm45686_stream.c#L205-L208

anthony289478 commented:
>The sensor_submit_fallback only kicks-in if the sensor driver doesn't implement the .submit API, which in this case it does:

I see this is implemented by the icm45686 as `icm45686_stream_submit` when streaming is enabled and `icm45686_submit_one_shot` if not enabled. In both cases it looks like `rtio_submit` is still called after reg read/write operations are queued and the process repeats. Eventually the jobs will need to go to the rtio_workq to be worked.

For this implementation, it will go to `spi_iodev_submit()`,

https://github.com/zephyrproject-rtos/zephyr/blob/d494f86ea40ef470ddaa2bb15ae83234345e7216/include/zephyr/drivers/spi.h#L1272-L1279

From there, the vendor implementation. In most cases, the default implementation is used,
```C
#ifdef CONFIG_SPI_RTIO
	.iodev_submit = spi_rtio_iodev_default_submit,
#endif
```
which ends up at the rtio_workq
https://github.com/zephyrproject-rtos/zephyr/blob/d494f86ea40ef470ddaa2bb15ae83234345e7216/drivers/spi/spi_rtio.c#L137-L153



> Take a look at:
> https://github.com/CogniPilot/zephyr/blob/a03c53eea78e4bb7c050177a938d214c2f6273d0/drivers/sensor/tdk/icm45686/icm45686_stream.c#L205-L208

It looks like that warning might exist to suppress the extra interrupts from watermark because the default configuration is set greater than or equals rather than just equals. 

I would say the warning should stay as a sanity check but the configuration should still be updated to only generate one interrupt per watermark.

Other TDK IMU's which have the same interrupt behavior have watermark interrupt only once per threshold  crossing so it is not an unusual request.

In the case of icm42688 wm_gt_th is disabled, and interrupts are not disabled in gpio callback because it is not needed.
https://github.com/zephyrproject-rtos/zephyr/blob/216fc8f5de132c967c4c846097e75258f6120f81/drivers/sensor/tdk/icm4268x/icm4268x_common.c#L315-L325
https://github.com/zephyrproject-rtos/zephyr/blob/216fc8f5de132c967c4c846097e75258f6120f81/drivers/sensor/tdk/icm4268x/icm4268x_trigger.c#L21-L37

on older drivers, such as icm42605, wm_gt_th is misconfigured as enabled and to suppress spurious interrupts, gpio interrupts are disabled in gpio callback.
https://github.com/zephyrproject-rtos/zephyr/blob/216fc8f5de132c967c4c846097e75258f6120f81/drivers/sensor/tdk/icm42605/icm42605_setup.c#L240
https://github.com/zephyrproject-rtos/zephyr/blob/216fc8f5de132c967c4c846097e75258f6120f81/drivers/sensor/tdk/icm42605/icm42605_trigger.c#L69

anthony289478 committed:
sensor: icm45686: Move watermark threshold mode to DT config

Move the watermark threshold trigger mode to a configurable dt boolean.

When using the default configuration of watermark threshold
interrupt greater than or equals, extra interrupts are serviced
to icm45686_event_handler().

When `fifo-watermark-equals;` is added to the sensor DT overlay,
the new behavior is only one interrupt is generated per watermark
threshold crossing. Until the host drains the fifo, no extra interrupts
will be generated.

Signed-off-by: Anthony Williams <anthony289478@gmail.com>

ubieda commented on drivers/sensor/tdk/icm45686/icm45686_stream.c:
```diff
@@ -152,7 +152,7 @@ static void icm45686_complete_handler(struct rtio *ctx,
 	if (should_flush_fifo(read_cfg, int_status)) {
 		uint8_t write_reg = REG_FIFO_CONFIG2_FIFO_FLUSH(true) |
-				    REG_FIFO_CONFIG2_FIFO_WM_GT_THS(true);
+				    REG_FIFO_CONFIG2_FIFO_WM_GT_THS(false);
```
Can we make this a dt-property? You'd be able to set it to change the behavior to be, say, fifo-wm-equals, and therefore the driver works for both use-cases.
